File: csharp/tests/TestStructCrudViaTable.cs
using System.Text.Json.Nodes;
using MyAirtable;
using Xunit;

namespace MyAirtable.Tests;

/// <summary>Dict-style (untyped) CRUD against the live base via <c>airtable.Primary.Dict</c>.</summary>
public class TestStructCrudViaTable
{
    private readonly Airtable _airtable = TestSetup.MakeAirtable();

    [Fact]
    public async Task PrimaryKeyOnlyCrudRoundTrip()
    {
        var primaryKey = TestSetup.PrimaryKey("StructCrud", "PKOnly");
        var fields = new Fields(new Dictionary<string, JsonNode?>(), PrimaryFields.NameToId);
        fields.SetString(PrimaryFields.PrimaryKeyId, primaryKey);

        var created = await _airtable.Primary.Dict.CreateAsync(fields);
        var recordId = created.Id;
        try
        {
            Assert.False(string.IsNullOrEmpty(created.Id));
            Assert.Equal(primaryKey, created.Fields.GetString(PrimaryFields.PrimaryKeyId));

            var fetched = await _airtable.Primary.Dict.GetAsync(recordId);
            Assert.Equal(recordId, fetched.Id);
            Assert.Equal(primaryKey, fetched.Fields.GetString(PrimaryFields.PrimaryKeyId));

            var update = new Fields(new Dictionary<string, JsonNode?>(), PrimaryFields.NameToId);
            var updatedKey = primaryKey + " Updated";
            update.SetString(PrimaryFields.PrimaryKeyId, updatedKey);
            var updated = await _airtable.Primary.Dict.UpdateAsync(recordId, update);
            Assert.Equal(updatedKey, updated.Fields.GetString(PrimaryFields.PrimaryKeyId));

            await _airtable.Primary.Dict.DeleteAsync(recordId);
            await Assert.ThrowsAnyAsync<AirtableException>(() =>
                _airtable.Primary.Dict.GetAsync(recordId)
            );
        }
        catch
        {
            await TryDeleteAsync(recordId);
            throw;
        }
    }

    [Fact]
    public async Task DictTableSupportsDualIdNameFieldAccess()
    {
        var primaryKey = TestSetup.PrimaryKey("StructCrud", "DualAccess");
        var fields = new Fields(new Dictionary<string, JsonNode?>(), PrimaryFields.NameToId);
        fields.SetString(PrimaryFields.PrimaryKeyId, primaryKey);

        var created = await _airtable.Primary.Dict.CreateAsync(fields);
        try
        {
            Assert.Equal(primaryKey, created.Fields.GetString(PrimaryFields.PrimaryKeyName));
            Assert.Equal(primaryKey, created.Fields.GetString(PrimaryFields.PrimaryKeyId));
        }
        finally
        {
            await TryDeleteAsync(created.Id);
        }
    }

    [Fact]
    public async Task AllSimplePropertiesRoundTrip()
    {
        var primaryKey = TestSetup.PrimaryKey("StructCrud", "AllProps");
        var fields = new Fields(new Dictionary<string, JsonNode?>(), PrimaryFields.NameToId);
        fields.SetString(PrimaryFields.PrimaryKeyId, primaryKey);
        fields.SetString(PrimaryFields.SingleLineTextId, "Hello World");
        fields.SetString(PrimaryFields.LongTextId, "Long text content");

        var created = await _airtable.Primary.Dict.CreateAsync(fields);
        try
        {
            var fetched = await _airtable.Primary.Dict.GetAsync(created.Id);
            Assert.Equal("Hello World", fetched.Fields.GetString(PrimaryFields.SingleLineTextId));
            Assert.Equal("Long text content", fetched.Fields.GetString(PrimaryFields.LongTextId));
        }
        finally
        {
            await TryDeleteAsync(created.Id);
        }
    }

    private async Task TryDeleteAsync(string recordId)
    {
        try
        {
            await _airtable.Primary.Dict.DeleteAsync(recordId);
        }
        catch (AirtableException)
        {
            // best-effort cleanup
        }
    }
}

File: cpp/tests/test_struct_crud_via_table.cpp
// Dict-style (untyped) CRUD against the live base via airtable.primary().dict()
// (C# TestStructCrudViaTable parity, 3 cases).
#include <catch2/catch_test_macros.hpp>

#include "test_setup.hpp"

using namespace myairtable;
using myairtable_tests::make_airtable;
using myairtable_tests::primary_key;

namespace {
void try_remove(Airtable& airtable, const std::string& record_id) {
    try {
        airtable.primary().dict().remove(record_id);
    } catch (const AirtableException&) {
        // best-effort cleanup
    }
}
} // namespace

TEST_CASE("struct crud: primary-key-only round trip", "[crud][struct-crud]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("StructCrud", "PKOnly");
    Fields fields({}, PrimaryFields::kNameToId);
    fields.set_string(PrimaryFields::kPrimaryKeyId, pk);

    const auto created = airtable.primary().dict().create(fields);
    const auto record_id = created.id;
    try {
        REQUIRE_FALSE(created.id.empty());
        REQUIRE(created.fields.get_string(PrimaryFields::kPrimaryKeyId) == pk);

        const auto fetched = airtable.primary().dict().get(record_id);
        REQUIRE(fetched.id == record_id);
        REQUIRE(fetched.fields.get_string(PrimaryFields::kPrimaryKeyId) == pk);

        Fields update({}, PrimaryFields::kNameToId);
        const auto updated_pk = pk + " Updated";
        update.set_string(PrimaryFields::kPrimaryKeyId, updated_pk);
        const auto updated = airtable.primary().dict().update(record_id, update);
        REQUIRE(updated.fields.get_string(PrimaryFields::kPrimaryKeyId) == updated_pk);

        airtable.primary().dict().remove(record_id);
        REQUIRE_THROWS_AS(airtable.primary().dict().get(record_id), AirtableException);
    } catch (...) {
        try_remove(airtable, record_id);
        throw;
    }
}

TEST_CASE("struct crud: dual id/name field access", "[crud][struct-crud]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("StructCrud", "DualAccess");
    Fields fields({}, PrimaryFields::kNameToId);
    fields.set_string(PrimaryFields::kPrimaryKeyId, pk);

    const auto created = airtable.primary().dict().create(fields);
    try {
        REQUIRE(created.fields.get_string(PrimaryFields::kPrimaryKeyName) == pk);
        REQUIRE(created.fields.get_string(PrimaryFields::kPrimaryKeyId) == pk);
    } catch (...) {
        try_remove(airtable, created.id);
        throw;
    }
    try_remove(airtable, created.id);
}

TEST_CASE("struct crud: all simple properties round trip", "[crud][struct-crud]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("StructCrud", "AllProps");
    Fields fields({}, PrimaryFields::kNameToId);
    fields.set_string(PrimaryFields::kPrimaryKeyId, pk);
    fields.set_string(PrimaryFields::kSingleLineTextId, "Hello World");
    fields.set_string(PrimaryFields::kLongTextId, "Long text content");

    const auto created = airtable.primary().dict().create(fields);
    try {
        const auto fetched = airtable.primary().dict().get(created.id);
        REQUIRE(fetched.fields.get_string(PrimaryFields::kSingleLineTextId) == "Hello World");
        REQUIRE(fetched.fields.get_string(PrimaryFields::kLongTextId) == "Long text content");
    } catch (...) {
        try_remove(airtable, created.id);
        throw;
    }
    try_remove(airtable, created.id);
}
